### PR TITLE
Complex support on GPU for dynamic casting

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Currently, the dynamic casting mechanism is implemented assuming no support of complex on GPU. This will no longer be true in the soon future.

https://github.com/pytorch/pytorch/pull/29547 could clear some clang warning but the complex support on GPU is still not complete:
- fetch is not supported
- casting between complex64 and complex128 is not supported
- complex scalar types are not tested

This PR is what should be done for type promotion in order to add support to complex dtype on GPU, as suggested in https://github.com/pytorch/pytorch/issues/755#issuecomment-552631381

Note that what is newly added here in this PR is not tested due to the lack of basic support of complex dtypes (I can not construct a complex tensor). But his PR shouldn't break any existing part of PyTorch. 

For the merge this PR, consider two options:
- We could merge this PR now so that @dylanbespalko could conveniently work based on master, if there is something wrong here not found by code review, @dylanbespalko would find when adding complex integration.
- Or, we could just leave this PR open, don't merge it. But then @dylanbespalko might need to manually apply this to his branch in order to support type promotion of complex.

### CODE REVIEW

File: c10/util/TypeCast.h [hunk 1]
```diff
@@ -103,16 +103,35 @@ C10_HOST_DEVICE inline dest_t static_cast_with_inter_type(src_t src) {
...
+  }
+};
+
+template<typename complex_src_t>
+struct MaybeReal<std::complex<double>, complex_src_t> {
+  static C10_HOST_DEVICE inline complex_src_t maybe_real(complex_src_t src) {
+    return src;
+  }
+};
+
```
Comment (dylanbespalko) [inline c10/util/TypeCast.h]: @zasdfgbnm 

Thanks for adding MaybeReal. I'm surprised that other scalar_t types like `Bool, Half, BFFloat16` can call the `scalar_t::real()` method, but I guess it worked for `std::real(scalar_t)`?

Comment (zasdfgbnm) [reply c10/util/TypeCast.h]: @dylanbespalko No, the way it is used makes sure that `complex_src_t` could only be `std::complex<something_t>`.

Comment (dylanbespalko) [reply c10/util/TypeCast.h]: Also, you could really help me out by adding this change to c10/util/TypeCast.h::line48::52. I think it should be:
```C++
template <typename dest_t, typename src_t>
C10_HOST_DEVICE inline dest_t static_cast_with_inter_type(src_t src) {
  return static_cast<dest_t>(
    static_cast<inter_copy_type_t<dest_t>>(MaybeReal<dest_t, src_t>(src)));
}
```

Comment (dylanbespalko) [reply c10/util/TypeCast.h]: Oh, I see. Clever...

**Comment by dylanbespalko** [comment]
@zasdfgbnm, @ezyang 

I would prefer if you accept these changes before mine. I will pull the update from master.

File: c10/util/TypeCast.h (hunk 1 above)
Comment (zasdfgbnm) [reply c10/util/TypeCast.h]: @dylanbespalko Thanks for the review. That makes great sense. And I will add that.

**Comment by zasdfgbnm** [comment]
@dylanbespalko I have moved logics into `static_cast_with_inter_type`. Could you please take another look and either approve or request change on this PR?

**Comment by dylanbespalko** [comment]
> @dylanbespalko I have moved logics into `static_cast_with_inter_type`. Could you please take another look and either approve or request change on this PR?

Looks good to me.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**